### DISCUSSION TITLE
uvm_analysis_imp class

### DISCUSSION THREAD

**Discussion by jiyong00** [pyuvm/pyuvm#151] [Ideas]
It may not be necessary in pyuvm, but I am used to the uvm_analysis_imp of systemverilog UVM.  So, I am using it as follows. What do you think pyuvm supports in this way?

```
from pyuvm import uvm_port_base

class uvm_analysis_imp(uvm_port_base):

    def __init__(self, name, parent, postfix=None):
        super().__init__(name, parent)

        self.name = name
        self.parent = parent
        self.postfix = postfix

    def write(self, item):
        func_name = 'write'
        if self.postfix:
            func_name += self.postfix # Use the 'postfix' instead of uvm_analysis_imp_decl macro
        getattr(self.parent, func_name)(item)

```
And, In the scoreboard,

```
class ScoreBoard(uvm_scoreboard):

    def __init__(self, name, parent):
        super().__init__(name, parent)
        self.yapp_port  = uvm_analysis_imp("yapp_port", self, "_yapp_port")

    def write_yapp_port(self, packet):
        # Make a copy for storing in the scoreboard
        sb_packet = packet.clone() # Clone returns uvm_object type

```

**Comment by raysalemi** [comment]
The UVM designers created the _imp classes to compensate for the fact that SystemVerilog does not support multiple inheritance. 

Could you describe your broader design challenge to help me understand how Python multiple inheritance could give you a clean solution?

For example, what if Scoreboard also extended uvm_analysis_export?

**Reply by jiyong00** [reply]
Actually, I didn't know that the _imp class was created to compensate for the fact that SystemVerilog doesn't support multiple inheritance :)
When creating multiple analysis interfaces within the scoreboard, do you think multiple inheritance can solve it?

**Reply by raysalemi** [reply]
Now, I am wondering what kind of problem you are trying to solve. If you are trying to get the Scoreboard to take data from two different sources, you can do it the way it's done in the example that comes with **pyuvm**. In this case

```
class Scoreboard(uvm_component):

    def build_phase(self):
        self.cmd_fifo = uvm_tlm_analysis_fifo("cmd_fifo", self)
        self.result_fifo = uvm_tlm_analysis_fifo("result_fifo", self)
        self.cmd_get_port = uvm_get_port("cmd_get_port", self)
        self.result_get_port = uvm_get_port("result_get_port", self)
        self.cmd_export = self.cmd_fifo.analysis_export
        self.result_export = self.result_fifo.analysis_export

    def connect_phase(self):
        self.cmd_get_port.connect(self.cmd_fifo.get_export)
        self.result_get_port.connect(self.result_fifo.get_export)

    def check_phase(self):
        passed = True
        try:
            self.errors = ConfigDB().get(self, "", "CREATE_ERRORS")
        except UVMConfigItemNotFound:
            self.errors = False
        while self.result_get_port.can_get():
            _, actual_result = self.result_get_port.try_get()
            cmd_success, cmd = self.cmd_get_port.try_get()
            if not cmd_success:
                self.logger.critical(f"result {actual_result} had no command")
            else:
                (A, B, op_numb) = cmd
                op = Ops(op_numb)
                predicted_result = alu_prediction(A, B, op, self.errors)
                if predicted_result == actual_result:
                    self.logger.info(f"PASSED: 0x{A:02x} {op.name} 0x{B:02x} ="
                                     f" 0x{actual_result:04x}")
                else:
                    self.logger.error(f"FAILED: 0x{A:02x} {op.name} 0x{B:02x} "
                                      f"= 0x{actual_result:04x} "
                                      f"expected 0x{predicted_result:04x}")
                    passed = False
        assert passed
```

The above gives you one scoreboard that can attach to multiple analysis ports.  

You can also extend `uvm_analysis_export` like this. There is no need for multiple_inheritance here since `uvm_analysis_export` extends `uvm_component`. 

```
# ## Extending the uvm_analysis_export class
# Figure 2: A component to add random numbers
class Adder(uvm_analysis_export):
    def start_of_simulation_phase(self):
        self.sum = 0

    def write(self, nn):
        self.sum += nn

    def report_phase(self):
        self.logger.info(f"Sum: {self.sum}")
```
This example is from the book in Chapter 36 and is here: 
https://github.com/raysalemi/Python4RTLVerification/blob/master/36_Analysis_ports/testbench.py

**Reply by jiyong00** [reply]
Hmm, I've always used uvm_analysis_imp among the various analysis interfaces in systemverilog UVM. However, it doesn't seem to exist in pyuvm. Yes, python does not support define macro, so I thought about how to apply `uvm_analysis_impl_decl in pyuvm. I just shared the idea, and it would be nice if pyuvm also made uvm_analysis_imp.

**Reply by raysalemi** [reply]
> Hmm, I've always used uvm_analysis_imp among the various analysis interfaces in systemverilog UVM. However, it doesn't seem to exist in pyuvm. Yes, python does not support define macro, so I thought about how to apply `uvm_analysis_impl_decl in pyuvm. I just shared the idea, and it would be nice if pyuvm also made uvm_analysis_imp.

There is no need for it in Python. The whole _imp mechanism is a workaround.

**Reply by suchi2108** [reply]
I have used the same to get data from two sources, referring https://github.com/raysalemi/Python4RTLVerification/blob/master/36_Analysis_ports/testbench.py. But somehow  I am getting the same data for both 'actual_result' and 'cmd' using 'try_get' for the loop. Although I logged datum in monitor, it's fine. Can you please guide me with this?

**Reply by raysalemi** [reply]
> I have used the same to get data from two sources, referring https://github.com/raysalemi/Python4RTLVerification/blob/master/36_Analysis_ports/testbench.py. But somehow  I am getting the same data for both 'actual_result' and 'cmd' using 'try_get' for the loop. Although I logged datum in monitor, it's fine. Can you please guide me with this?

Do you have code to share?

**Reply by suchi2108** [reply]
Sure.

**#Monitor**



    def __init__(self, name, parent, method_name):
        super().__init__(name, parent)
        self.method_name = method_name

    def build_phase(self):
        self.ap = uvm_analysis_port("ap", self)
        self.bfm = ConfigDB().get(self, "", "BFM")
        self.get_method = getattr(self.bfm, self.method_name)

    async def run_phase(self):
        while True:
            datum = await self.get_method()
            self.ap.write(datum)
            self.logger.debug(f"MONITORED {datum}")

**#Scoreboard**


    def build_phase(self):
        self.cmd_fifo = uvm_tlm_analysis_fifo("cmd_fifo", self)
        self.result_fifo = uvm_tlm_analysis_fifo("result_fifo", self)
        self.cmd_get_port = uvm_get_port("cmd_get_port", self)
        self.result_get_port = uvm_get_port("result_get_port", self)
        self.cmd_export = self.cmd_fifo.analysis_export
        self.result_export = self.result_fifo.analysis_export

    def connect_phase(self):
        self.cmd_get_port.connect(self.cmd_fifo.get_export)
        self.result_get_port.connect(self.result_fifo.get_export)

    def check_phase(self):
        while self.result_get_port.can_get():
            _, actual_result = self.result_get_port.try_get()
            cmd_success, cmd = self.cmd_get_port.try_get()
            if not cmd_success:
                self.logger.critical(f"result {actual_result} had no command")
            else:
                (A, B, op_numb) = cmd
                op = Ops(op_numb.value.integer)
                predicted_result = alu_ref(A.value.integer, B.value.integer, op)
                if predicted_result == actual_result.value.integer:
                    self.logger.info(f"PASSED: A is {hex(A.value.integer)}, Op is {op.name}, and B is {hex(B.value.integer)}"
                                     f" Actual result is {hex(actual_result.value.integer)}")
                else:
                    self.logger.error(f"FAILED: A is {A.value.integer}, Op is {op.name} and B is {B.value.integer} "
                                      f" Actual result is {actual_result.value.integer} "
                                      f"Expected result is {predicted_result}")


**Log:**
 0.00ns INFO     cocotb                             Running on xmsim(64) version 23.03-s004
     0.00ns INFO     cocotb                             Running tests with cocotb v1.7.1 from /.local/lib/python3.6/site-packages/cocotb
     0.00ns INFO     cocotb                             Seeding Python random module with 1691007910
     0.00ns INFO     cocotb.regression                  Found test alu_test.AluTest
     0.00ns INFO     cocotb.regression                  running AluTest (1/1)
     0.00ns DEBUG    ..st_top.env.driver140135958310192 alu_driver.py(15)[uvm_test_top.env.driver]: Sent command: cmd_tr : A: -3 OP: ADD (0) B: 3
     0.00ns DEBUG    ..st_top.env.driver140135958310192 alu_driver.py(15)[uvm_test_top.env.driver]: Sent command: cmd_tr : A: -5 OP: SUB (1) B: -7
     0.00ns DEBUG    ..t_top.env.cmd_mon140135958347960 alu_monitor.py(18)[uvm_test_top.env.cmd_mon]: MONITORED (ModifiableObject(alu.a), ModifiableObject(alu.b), ModifiableObject(alu.op))
     0.00ns DEBUG    ..op.env.result_mon140135958348352 alu_monitor.py(18)[uvm_test_top.env.result_mon]: MONITORED xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
    10.00ns DEBUG    ..st_top.env.driver140135958310192 alu_driver.py(15)[uvm_test_top.env.driver]: Sent command: cmd_tr : A: 6 OP: NOT (2) B: 10
    10.00ns DEBUG    ..t_top.env.cmd_mon140135958347960 alu_monitor.py(18)[uvm_test_top.env.cmd_mon]: MONITORED (ModifiableObject(alu.a), ModifiableObject(alu.b), ModifiableObject(alu.op))
    10.00ns DEBUG    ..op.env.result_mon140135958348352 alu_monitor.py(18)[uvm_test_top.env.result_mon]: MONITORED xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
    20.00ns DEBUG    ..st_top.env.driver140135958310192 alu_driver.py(15)[uvm_test_top.env.driver]: Sent command: cmd_tr : A: 9 OP: NAND (3) B: 4
    20.00ns DEBUG    ..t_top.env.cmd_mon140135958347960 alu_monitor.py(18)[uvm_test_top.env.cmd_mon]: MONITORED (ModifiableObject(alu.a), ModifiableObject(alu.b), ModifiableObject(alu.op))
    20.00ns DEBUG    ..op.env.result_mon140135958348352 alu_monitor.py(18)[uvm_test_top.env.result_mon]: MONITORED 00000000000000000000000000000000
    30.00ns DEBUG    ..st_top.env.driver140135958310192 alu_driver.py(15)[uvm_test_top.env.driver]: Sent command: cmd_tr : A: -9 OP: NOR (4) B: -6
    30.00ns DEBUG    ..t_top.env.cmd_mon140135958347960 alu_monitor.py(18)[uvm_test_top.env.cmd_mon]: MONITORED (ModifiableObject(alu.a), ModifiableObject(alu.b), ModifiableObject(alu.op))
    30.00ns DEBUG    ..op.env.result_mon140135958348352 alu_monitor.py(18)[uvm_test_top.env.result_mon]: MONITORED 00000000000000000000000000000010
    40.00ns DEBUG    ..st_top.env.driver140135958310192 alu_driver.py(15)[uvm_test_top.env.driver]: Sent command: cmd_tr : A: -10 OP: AND (5) B: 6
    40.00ns DEBUG    ..t_top.env.cmd_mon140135958347960 alu_monitor.py(18)[uvm_test_top.env.cmd_mon]: MONITORED (ModifiableObject(alu.a), ModifiableObject(alu.b), ModifiableObject(alu.op))
    40.00ns DEBUG    ..op.env.result_mon140135958348352 alu_monitor.py(18)[uvm_test_top.env.result_mon]: MONITORED 11111111111111111111111111111001
    50.00ns DEBUG    ..st_top.env.driver140135958310192 alu_driver.py(15)[uvm_test_top.env.driver]: Sent command: cmd_tr : A: 10 OP: OR (6) B: 9
    50.00ns DEBUG    ..t_top.env.cmd_mon140135958347960 alu_monitor.py(18)[uvm_test_top.env.cmd_mon]: MONITORED (ModifiableObject(alu.a), ModifiableObject(alu.b), ModifiableObject(alu.op))
    50.00ns DEBUG    ..op.env.result_mon140135958348352 alu_monitor.py(18)[uvm_test_top.env.result_mon]: MONITORED 11111111111111111111111111111111
    60.00ns DEBUG    ..st_top.env.driver140135958310192 alu_driver.py(15)[uvm_test_top.env.driver]: Sent command: cmd_tr : A: 0 OP: XOR (7) B: -9
    60.00ns DEBUG    ..t_top.env.cmd_mon140135958347960 alu_monitor.py(18)[uvm_test_top.env.cmd_mon]: MONITORED (ModifiableObject(alu.a), ModifiableObject(alu.b), ModifiableObject(alu.op))
    60.00ns DEBUG    ..op.env.result_mon140135958348352 alu_monitor.py(18)[uvm_test_top.env.result_mon]: MONITORED 00000000000000000000000000000000
    70.00ns DEBUG    ..t_top.env.cmd_mon140135958347960 alu_monitor.py(18)[uvm_test_top.env.cmd_mon]: MONITORED (ModifiableObject(alu.a), ModifiableObject(alu.b), ModifiableObject(alu.op))
    70.00ns DEBUG    ..op.env.result_mon140135958348352 alu_monitor.py(18)[uvm_test_top.env.result_mon]: MONITORED 00000000000000000000000000000110
    80.00ns DEBUG    ..t_top.env.cmd_mon140135958347960 alu_monitor.py(18)[uvm_test_top.env.cmd_mon]: MONITORED (ModifiableObject(alu.a), ModifiableObject(alu.b), ModifiableObject(alu.op))
    80.00ns DEBUG    ..op.env.result_mon140135958348352 alu_monitor.py(18)[uvm_test_top.env.result_mon]: MONITORED 00000000000000000000000000001011
    90.00ns DEBUG    ..t_top.env.cmd_mon140135958347960 alu_monitor.py(18)[uvm_test_top.env.cmd_mon]: MONITORED (ModifiableObject(alu.a), ModifiableObject(alu.b), ModifiableObject(alu.op))
    90.00ns DEBUG    ..op.env.result_mon140135958348352 alu_monitor.py(18)[uvm_test_top.env.result_mon]: MONITORED 11111111111111111111111111110111
    90.00ns ERROR    .._top.env.coverage140135958348744 alu_coverage.py(17)[uvm_test_top.env.coverage]: Functional coverage error. Missed: {<Ops.ADD: 0>, <Ops.SUB: 1>, <Ops.NOT: 2>, <Ops.NAND: 3>, <Ops.NOR: 4>, <Ops.AND: 5>, <Ops.OR: 6>, <Ops.XOR: 7>}
    90.00ns INFO     ..op.env.scoreboard140135958349584 alu_scoreboard.py(30)[uvm_test_top.env.scoreboard]: PASSED: A is 0x0, Op is XOR, and B is 0xfffffff7 Actual result is 0xfffffff7
    90.00ns INFO     ..op.env.scoreboard140135958349584 alu_scoreboard.py(30)[uvm_test_top.env.scoreboard]: PASSED: A is 0x0, Op is XOR, and B is 0xfffffff7 Actual result is 0xfffffff7
    90.00ns INFO     ..op.env.scoreboard140135958349584 alu_scoreboard.py(30)[uvm_test_top.env.scoreboard]: PASSED: A is 0x0, Op is XOR, and B is 0xfffffff7 Actual result is 0xfffffff7
    90.00ns INFO     ..op.env.scoreboard140135958349584 alu_scoreboard.py(30)[uvm_test_top.env.scoreboard]: PASSED: A is 0x0, Op is XOR, and B is 0xfffffff7 Actual result is 0xfffffff7
    90.00ns INFO     ..op.env.scoreboard140135958349584 alu_scoreboard.py(30)[uvm_test_top.env.scoreboard]: PASSED: A is 0x0, Op is XOR, and B is 0xfffffff7 Actual result is 0xfffffff7
    90.00ns INFO     ..op.env.scoreboard140135958349584 alu_scoreboard.py(30)[uvm_test_top.env.scoreboard]: PASSED: A is 0x0, Op is XOR, and B is 0xfffffff7 Actual result is 0xfffffff7
    90.00ns INFO     ..op.env.scoreboard140135958349584 alu_scoreboard.py(30)[uvm_test_top.env.scoreboard]: PASSED: A is 0x0, Op is XOR, and B is 0xfffffff7 Actual result is 0xfffffff7
    90.00ns INFO     ..op.env.scoreboard140135958349584 alu_scoreboard.py(30)[uvm_test_top.env.scoreboard]: PASSED: A is 0x0, Op is XOR, and B is 0xfffffff7 Actual result is 0xfffffff7
    90.00ns INFO     ..op.env.scoreboard140135958349584 alu_scoreboard.py(30)[uvm_test_top.env.scoreboard]: PASSED: A is 0x0, Op is XOR, and B is 0xfffffff7 Actual result is 0xfffffff7
    90.00ns INFO     ..op.env.scoreboard140135958349584 alu_scoreboard.py(30)[uvm_test_top.env.scoreboard]: PASSED: A is 0x0, Op is XOR, and B is 0xfffffff7 Actual result is 0xfffffff7
    90.00ns INFO     cocotb.regression                  AluTest passed
    90.00ns INFO     cocotb.regression                  **************************************************************************************
                                                        ** TEST                          STATUS  SIM TIME (ns)  REAL TIME (s)  RATIO (ns/s) **
                                                        **************************************************************************************
                                                        ** alu_test.AluTest               PASS          90.00           0.02       5995.83  **
                                                        **************************************************************************************
                                                        ** TESTS=1 PASS=1 FAIL=0 SKIP=0                 90.00           0.95         94.82  **
                                                        **************************************************************************************


DUT is ALU with  signals a, b, op, and r. Please let me know if anything else is required.

**Reply by raysalemi** [reply]
Could you share the component that connects the monitor and the scoreboard?

**Reply by suchi2108** [reply]
They are connected in Environment.

#Environment

    def build_phase(self):
        self.seqr = uvm_sequencer("seqr", self)
        ConfigDB().set(None, "*", "SEQR", self.seqr)
        self.driver = Driver("driver", self)
        self.cmd_mon = Monitor("cmd_mon", self, "get_cmd")
        self.result_mon = Monitor("result_mon", self, "get_result")
        self.coverage = Coverage("coverage", self)
        self.scoreboard = Scoreboard("scoreboard", self)

    def connect_phase(self):
        self.driver.seq_item_port.connect(self.seqr.seq_item_export)
        self.cmd_mon.ap.connect(self.scoreboard.cmd_export)
        self.result_mon.ap.connect(self.scoreboard.result_export)
        self.cmd_mon.ap.connect(self.coverage.analysis_export)

**Comment by raysalemi** [comment]
Also, could you share only the output line that shows the error with a few on either side for context?

**Reply by suchi2108** [reply]
Log from monitor (Couldn't find a way to convert datum here for cmd)
![image](https://github.com/pyuvm/pyuvm/assets/118723121/8b60e853-3b0f-4c4b-b625-bc8e720355cf)


Log from scoreboard just before the checker function

![image](https://github.com/pyuvm/pyuvm/assets/118723121/0fe39f9d-dada-4333-b595-2100555600ef)

**Reply by raysalemi** [reply]
Above, you said you were getting the same for `actual_result` and `cmd`.  But you seem to be getting a command and actual result separately.  

What is the bug again?

**Reply by suchi2108** [reply]
I meant, as the driver is sending 8 transactions for different operations of ALU. In the FIFOs they should contain 8 different cmd and actual_result respectively. Instead, it has same in both FIFOs (can be seen in log from scoreboard). I hope I clear what my query is.

**Reply by c-bjerrum** [reply]
Hello, it seems to me that you are sending the signals and not their values from your monitor, so when you then read all the values at 90 ns, they are all the same as you are reading the signal value at this time . 

I think a way of solving your problem would be to access the current values in your monitor as:
```
 (A, B, op_numb) = datum
 new_datum = (A.value, B.value, op_numb.value) 
 self.ap.write(new_datum) 
```

**Reply by raysalemi** [reply]
Aha!  That explains why they are all identical handles.

**Reply by suchi2108** [reply]
Yes that was the case, I could fix it. Thank you :)

**Comment by raysalemi** [comment]
My gut says that you are sending the same command handle each time. I wonder if a variable is not getting overwritten with a new transaction somewhere.
